File: solution/203RemoveLinkedListElements/203RemoveLinkedListElements.cpp
#include <iostream>
using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
  ListNode *removeElements(ListNode *head, int val) {
    ListNode dummy(0);
    dummy.next = head;
    ListNode *r = &dummy;
    ListNode *p = head;
    while (p != nullptr) {
      if (p->val == val) {
        r->next = p->next;
      } else {
        r = p;
      }
      p = p->next;
    }
    return dummy.next;
  }
};

int main() {}

File: src/203RemoveLinkedListElements/203RemoveLinkedListElements.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_elements(head: Option<Box<ListNode>>, val: i32) -> Option<Box<ListNode>> {
        let mut dummy = ListNode{val:0, next:head};
        let mut p = &mut dummy;
        while let Some(q) = p.next.as_mut() {
            if q.val == val {
                p.next = q.next.take();
            }else{
                p = p.next.as_mut().unwrap();
            }
        }
        
        dummy.next
    }
}
